src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs

```diff
@@ -23,47 +22,100 @@ internal partial class Checksum
         // https://github.com/dotnet/runtime/blob/f2db6d6093c54e5eeb9db2d8dcbe15b2db92ad8c/src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/SHA256.cs#L18-L19
         private const int SHA256HashSizeBytes = 256 / 8;
 
+#if NET5_0_OR_GREATER
         private static readonly ObjectPool<IncrementalHash> s_incrementalHashPool =
             new(() => IncrementalHash.CreateHash(HashAlgorithmName.SHA256), size: 20);
+#else
+        private static readonly ObjectPool<SHA256> s_incrementalHashPool =
+            new(SHA256.Create, size: 20);
+#endif
 
+#if !NET5_0_OR_GREATER
         // Dedicated pools for the byte[]s we use to create checksums from two or three existing checksums. Sized to
         // exactly the space needed to splat the existing checksum data into the array and then hash it.
 
         private static readonly ObjectPool<byte[]> s_twoChecksumByteArrayPool = new(() => new byte[HashSize * 2]);
         private static readonly ObjectPool<byte[]> s_threeChecksumByteArrayPool = new(() => new byte[HashSize * 3]);
+#endif
 
         public static Checksum Create(IEnumerable<string> values)
         {
+#if NET5_0_OR_GREATER
+            using var pooledHash = s_incrementalHashPool.GetPooledObject();
+
+            foreach (var value in values)
+            {
+                pooledHash.Object.AppendData(MemoryMarshal.AsBytes(value.AsSpan()));
+                pooledHash.Object.AppendData(MemoryMarshal.AsBytes("\0".AsSpan()));
+            }
+
+            Span<byte> hash = stackalloc byte[SHA256HashSizeBytes];
+            pooledHash.Object.GetHashAndReset(hash);
+            return From(hash);
+#else
             using var pooledHash = s_incrementalHashPool.GetPooledObject();
             using var pooledBuffer = SharedPools.ByteArray.GetPooledObject();
             var hash = pooledHash.Object;
 
+            hash.Initialize();
             foreach (var value in values)
             {
                 AppendData(hash, pooledBuffer.Object, value);
                 AppendData(hash, pooledBuffer.Object, "\0");
             }
 
-            return From(hash.GetHashAndReset());
+            return From(hash.TransformFinalBlock(Array.Empty<byte>(), 0, 0));
+#endif
         }
 
         public static Checksum Create(string value)
         {
+#if NET5_0_OR_GREATER
+            Span<byte> hash = stackalloc byte[SHA256HashSizeBytes];
+            SHA256.HashData(MemoryMarshal.AsBytes(value.AsSpan()), hash);
+            return From(hash);
+#else
             using var pooledHash = s_incrementalHashPool.GetPooledObject();
             using var pooledBuffer = SharedPools.ByteArray.GetPooledObject();
             var hash = pooledHash.Object;
+            hash.Initialize();
 
             AppendData(hash, pooledBuffer.Object, value);
 
-            return From(hash.GetHashAndReset());
+            return From(hash.TransformFinalBlock(Array.Empty<byte>(), 0, 0));
```

Comment:
what's the change about 'Initialize/TransformFinalBlock' about?

Reply:
➡️ We change from `IncrementalHash` (which uses `SHA256CryptoServiceProvider` internally) to using `SHA256` directly (via `SHA256.Create()`). `Initialize` and `TransformFinalBlock` are what `IncrementalHash` uses internally.